### DISCUSSION TITLE
Glass material / View distortion to fake refraction

### DISCUSSION THREAD

**Discussion by Kolupsy** [bnpr/Malt#179] [Q&A] [answered]
How would one make a glass like shader or a fake refraction shader in MALT? My first idea was that you would need to do this on a screen shader but the test that I had in mind did not really work out. The plan was to pass the matcap UVs of the sphere (the object that would distort the image behind it) to a custom output and have it distort the Screen UVs for the rendered image.
This was my attempt but all it did was make the sphere appear a little smaller and more oval shaped.
![image](https://user-images.githubusercontent.com/88381426/143584369-bc27f52c-3e64-4985-bbe0-1f4d05cae40c.png)

**Comment by pragma37** [comment]
In the built-in NPR_Pipeline, transparency is achieved through Depth Peeling layers (rendered front to back).
https://github.com/bnpr/Malt/tree/Development/Malt/Pipelines/NPR_Pipeline#render-overview
So the Color from your Render Layer Input only has the color from the current transparent layer being rendered.

Refraction is a common feature.  I have to add a way to (at least) read the Color result of the Opaque Layer from the Transparent Layers, but  at the moment that's not possible to achieve without rolling your own custom pipeline.

**Reply by Kolupsy** [reply]
hm i thought i was a tweet from you or light a whole while back showing off a refraction kind of setup in Malt. But thanks for the input. Would be cool to get some image distortion happening in malt.

**Comment by Kolupsy** [comment, marked as answer]
Fake Refraction now possible:
![image](https://user-images.githubusercontent.com/88381426/147364570-ee4e7bd9-1a4a-455b-835e-e26132f6beb0.png)

Link: https://github.com/bnpr/Malt/commit/9fbf6e49b939a40d7f0bdc9abd1ac12f069734c1

Shader function:
```
vec4 sample_opaque(vec2 uv)
{
    #ifdef PIXEL_SHADER
    {
        if(Settings.Transparency)
        {
            return texture(IN_OPAQUE_COLOR, uv);
        }
    }
    #endif
    return vec4(0);
}
```

**Comment by Kolupsy** [comment]
what is the current status on fake refraction. IIRC you wanted to find a better way to implement this. Not sure if I understand depth peeling correctly would it be possible to sample not only the opaque color but simply the previous transparent layer to make 'double refraction' possible?

**Reply by pragma37** [reply]
There have been no changes in that regard.

The problem is that transparent layers are currently rendered front to back.
The "easiest" solution would be to render all the Pre-Pass layers first, so then the Main-Pass can go back to front by comparing against the Pre-Pass depths.

Ideally, we should use a per-pixel linked list for the Pre-Pass, so transparent meshes could be rendered only once instead of once for each layer.
And we could do a similar optimization for the Main Pass in the cases where refraction is not needed.

But I have been so focused on the UI side that I barely touched actual rendering in a while.
